### ISSUE TITLE
availability-distribution: potential memory leak

### ISSUE THREAD

**Issue by sandreim** [paritytech/polkadot-sdk#5258]
This is a nice pattern of increased CPU usage over time. I didn't look much at it, but it might be some iteration of a collection that doesn't get cleaned up.

<img width="1401" alt="Screenshot 2024-08-06 at 14 35 31" src="https://github.com/user-attachments/assets/2c1ee419-1638-4827-b186-2a4646c0a765">

https://grafana.teleport.parity.io/goto/9zQcAE9IR?orgId=1

CC @alindima

**Comment by alindima** [comment]
I looked into this.

The subsystem itself is not doing any work in the main loop, as it spawns separate tasks for everything.

The CPU usage for the tasks that the subsystem spawns (pov-receiver, chunk-receiver, pov-fetcher, chunk-fetcher) all look normal.

I discovered the culprit: Jaeger spans.
We hold a local hashmap of spans: https://github.com/paritytech/polkadot-sdk/blob/ebcbca3ff606b22b5eb81bcbfaa9309752d64dde/polkadot/node/network/availability-distribution/src/lib.rs#L107
which adds a new entry on every leaf activation, but only gets cleaned up when a block is finalized:

https://github.com/paritytech/polkadot-sdk/blob/ebcbca3ff606b22b5eb81bcbfaa9309752d64dde/polkadot/node/network/availability-distribution/src/lib.rs#L175

However, when a block gets finalized, we only get notified about the block with the largest height (the leaf), so we never clean up spans for relay parents that got finalized in a batch (since Grandpa finalizes chains, not blocks). Therefore, the size of the hashmap is almost always strictly increasing.

Now, to fix this there are a couple of options:
1. Remove Jaeger altoghether: https://github.com/paritytech/polkadot-sdk/issues/4995
2. Fix this somehow. Simplest idea is to also store the block number for each span and when getting a finalized block, also trim all spans with lower block heights. Another option is to modify the overseer to also supply the active leaves chain that have been deactivated by this finalization.
